rbs/prism/SigsRewriterPrism.cc

```diff
@@ -1,17 +1,770 @@
 #include "rbs/prism/SigsRewriterPrism.h"
 
+#include "absl/strings/match.h"
+#include "absl/strings/str_split.h"
+#include "core/errors/rewriter.h"
+#include "parser/helper.h"
+#include "parser/prism/Helpers.h"
+#include "rbs/SignatureTranslator.h"
+#include "rbs/TypeParamsToParserNodes.h"
+
+extern "C" {
+#include "prism.h"
+}
+#include "rbs/prism/SignatureTranslatorPrism.h"
+
 using namespace std;
+using namespace sorbet::parser::Prism;
 
 namespace sorbet::rbs {
 
-SigsRewriterPrism::SigsRewriterPrism(core::MutableContext ctx, const parser::Prism::Parser &parser,
-                                     std::map<pm_node_t *, std::vector<CommentNodePrism>> &commentsByNode)
-    : ctx(ctx), parser(parser), commentsByNode(&commentsByNode) {}
+namespace {
+
+pm_node_t *signaturesTarget(pm_node_t *node, const parser::Prism::Parser &parser) {
+    if (node == nullptr) {
+        return nullptr;
+    }
+
+    switch (PM_NODE_TYPE(node)) {
+        case PM_DEF_NODE:
+            // Both instance methods and singleton methods use PM_DEF_NODE in Prism
+            // (singleton methods have a receiver field set)
+            return node;
+        case PM_CALL_NODE: {
+            if (parser.isAttrAccessorCall(node) || parser.isVisibilityCall(node)) {
+                return node;
+            }
+            return nullptr;
+        }
+        default:
+            return nullptr;
+    }
+}
+
+/**
+ * Extracts and parses the argument from the annotation string.
+ *
+ * Considering an annotation like `@requires_ancestor: SomeModule`,
+ * this function will extract and parse `SomeModule` as a type then return the corresponding Prism node.
+ *
+ * We do not error if the node is not a constant, we just insert it as is and let the pipeline error down the line.
+ */
+pm_node_t *extractHelperArgument(core::MutableContext ctx, parser::Prism::Parser &parser, const Comment &annotation,
+                                 int offset) {
+    while (annotation.string[offset] == ' ') {
+        offset++;
+    }
+
+    Comment comment = {
+        core::LocOffsets{annotation.typeLoc.beginPos() + offset, annotation.typeLoc.endPos()},
+        core::LocOffsets{annotation.typeLoc.beginPos() + offset, annotation.typeLoc.endPos()},
+        annotation.string.substr(offset),
+    };
+
+    return rbs::SignatureTranslatorPrism(ctx, parser).translateType(RBSDeclaration{vector<Comment>{comment}});
+}
+
+/**
+ * Extracts class-level helper method calls from annotations.
+ *
+ * For example, given the annotations:
+ *     # @abstract
+ *     # @interface
+ *
+ * This function will return Prism call nodes for:
+ *     self.abstract!()
+ *     self.interface!()
+ *
+ * It doesn't insert them into the body of the class/module/etc.
+ */
+vector<pm_node_t *> extractHelpers(core::MutableContext ctx, absl::Span<const Comment> annotations,
+                                   parser::Prism::Parser &parser) {
+    Factory prism(parser);
```

Comment:
The linter doesn't strictly require it, but try to use uniform initialization sytnax whenever you can


```suggestion
    Factory prism{parser};
```